File: Programmers/Python/더 맵게.py
# 우선순위큐라고도 불리는 heapq는 Index 0에 대해서 가장 작은 값은 반환한다는 특징을 가짐

import heapq
def solution(scoville, K):
    answer = 0
    # 원래 선언된 리스트에 대하여 heap구조로 변환하는 함수 : heapify
    heapq.heapify(scoville)

    while len(scoville) > 0:
        sum = 0
        if scoville[0] < K and len(scoville) >= 2:
            sum += scoville[0]
            heapq.heappop(scoville)  # heap 내에서 가장 작은 값을 반환
            sum += scoville[0] * 2
            heapq.heappop(scoville)
            heapq.heappush(scoville, sum)
            answer += 1
            
        elif len(scoville) == 1 and scoville[0] < K:
            return -1
        
        else:
            heapq.heappop(scoville)
    
    return answer

File: Programmers/C++/더 맵게.cpp
#include <vector>
#include <queue>

using namespace std;

int solution(vector<int> scoville, int K) {
    int answer = 0;
    priority_queue <int, vector<int>, greater<>> pq;
    for (int &i : scoville){ pq.push(i); }
    
    while (pq.size() > 0){
        int sum = 0;
        // K보다 작으면 합치기
        if (pq.top() < K and pq.size() >= 2){
            // 가장 작은 값과 그 다음 작은 값을 합침
            sum += pq.top();
            pq.pop();  // pq 내에서 가장 작은 값을 반환
            sum += pq.top() * 2;
            pq.pop();
            
            pq.push(sum);
            answer += 1;
        }
        // 모든 음식의 스코빌 지수를 K 이상으로 만들 수 없는 경우    
        else if (pq.size() == 1 and pq.top() < K){
            return -1;
        }
        // K 보다 크면 지우기
        else{
            pq.pop();
        }
    }
                
    return answer;
}
